pkg/dogstatsd/batch.go

```diff
@@ -9,44 +9,95 @@ import (
 )
 
 // batcher batches multiple metrics before submission
 // this struct is not safe for concurrent use
 type batcher struct {
-	samples      []metrics.MetricSample
-	samplesCount int
+	samples      [][]metrics.MetricSample
+	samplesCount []int
 
 	events        []*metrics.Event
 	serviceChecks []*metrics.ServiceCheck
 
 	// output channels
-	choutSamples       chan<- []metrics.MetricSample
 	choutEvents        chan<- []*metrics.Event
 	choutServiceChecks chan<- []*metrics.ServiceCheck
 
 	metricSamplePool *metrics.MetricSamplePool
+
+	demux aggregator.Demultiplexer
+	// buffer slice allocated once per contextResolver to combine and sort
+	// tags, origin detection tags and k8s tags.
+	tagsBuffer    *tagset.HashingTagsAccumulator
+	keyGenerator  *ckey.KeyGenerator
+	pipelineCount int
+}
+
+// Use fastrange instead of a modulo for better performances.
+// See http://lemire.me/blog/2016/06/27/a-fast-alternative-to-the-modulo-reduction/.
+//
+// Note that we shift the context key because it is an actual 64 bits, and
+// the fast range has to operate on 32 bits values, so, we shift it in order
+// to "reduce" its size to 32 bits, we don't mind using only half of the context
+// key for the shard key, it will be unique enough for such a purpose.
+func fastrange(key ckey.ContextKey, pipelineCount int) uint32 {
+	// return uint32(uint64(key) % uint64(pipelineCount))
+	return uint32((uint64(key>>32) * uint64(pipelineCount)) >> 32)
 }
 
 func newBatcher(demux aggregator.Demultiplexer) *batcher {
 	agg := demux.Aggregator()
-	s, e, sc := agg.GetBufferedChannels()
+
+	e, sc := agg.GetBufferedChannels()
+	pipelineCount := config.Datadog.GetInt("dogstatsd_pipeline_count")
+	if pipelineCount <= 0 {
+		pipelineCount = 1
+	}
+	samples := make([][]metrics.MetricSample, pipelineCount)
+	samplesCount := make([]int, pipelineCount)
+
+	for i := range samples {
+		samples[i] = demux.GetMetricSamplePool().GetBatch()
+		samplesCount[i] = 0
+	}
+
 	return &batcher{
-		samples:            agg.MetricSamplePool.GetBatch(),
-		metricSamplePool:   agg.MetricSamplePool,
-		choutSamples:       s,
+		samples:            samples,
+		samplesCount:       samplesCount,
+		metricSamplePool:   demux.GetMetricSamplePool(),
 		choutEvents:        e,
 		choutServiceChecks: sc,
+
+		demux:         demux,
+		pipelineCount: pipelineCount,
+		tagsBuffer:    tagset.NewHashingTagsAccumulator(),
+		keyGenerator:  ckey.NewKeyGenerator(),
 	}
 }
 
 func (b *batcher) appendSample(sample metrics.MetricSample) {
-	if b.samplesCount == len(b.samples) {
-		b.flushSamples()
+	var shardKey uint32
+	if b.pipelineCount > 1 {
+		// TODO(remy): re-using this tagsBuffer later in the pipeline (by sharing
+		// it in the sample?) would reduce CPU usage, avoiding to recompute
+		// the tags hashes while generating the context key.
+		b.tagsBuffer.Append(sample.Tags...)
+		h := b.keyGenerator.Generate(sample.Name, sample.Host, b.tagsBuffer)
+		b.tagsBuffer.Reset()
+		shardKey = fastrange(h, b.pipelineCount)
```

Comment:
Can a round robin algorithm be used instead?

Reply:
No, it would lead to incorrect sampling of the data.